File: project_euler/problem_215/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int W = 32;
int H = 10;

vector<vector<int>> rows; // each row is a set of crack positions

void generate(int pos, vector<int>& cracks) {
    if (pos == W) {
        rows.push_back(cracks);
        return;
    }
    // Try placing a 2-brick
    if (pos + 2 <= W) {
        if (pos + 2 < W) cracks.push_back(pos + 2);
        generate(pos + 2, cracks);
        if (pos + 2 < W) cracks.pop_back();
    }
    // Try placing a 3-brick
    if (pos + 3 <= W) {
        if (pos + 3 < W) cracks.push_back(pos + 3);
        generate(pos + 3, cracks);
        if (pos + 3 < W) cracks.pop_back();
    }
}

int main() {
    vector<int> cracks;
    generate(0, cracks);

    int m = rows.size();

    // Convert crack positions to bitmask for fast compatibility check
    vector<unsigned int> mask(m, 0);
    for (int i = 0; i < m; i++) {
        for (int c : rows[i]) {
            mask[i] |= (1u << c);
        }
    }

    // Build compatibility list
    vector<vector<int>> compat(m);
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < m; j++) {
            if ((mask[i] & mask[j]) == 0) {
                compat[i].push_back(j);
            }
        }
    }

    // DP
    vector<long long> dp(m, 1); // height 1: each row has 1 way

    for (int h = 2; h <= H; h++) {
        vector<long long> ndp(m, 0);
        for (int i = 0; i < m; i++) {
            for (int j : compat[i]) {
                ndp[i] += dp[j];
            }
        }
        dp = ndp;
    }

    long long answer = 0;
    for (int i = 0; i < m; i++) {
        answer += dp[i];
    }

    cout << answer << endl;
    return 0;
}

File: project_euler/problem_215/solution.py
def solve():
    W = 32
    H = 10

    # Generate all row tilings as sets of crack positions
    rows = []

    def generate(pos, cracks):
        if pos == W:
            rows.append(frozenset(cracks))
            return
        # Try 2-brick
        if pos + 2 <= W:
            new_cracks = cracks + ([pos + 2] if pos + 2 < W else [])
            generate(pos + 2, new_cracks)
        # Try 3-brick
        if pos + 3 <= W:
            new_cracks = cracks + ([pos + 3] if pos + 3 < W else [])
            generate(pos + 3, new_cracks)

    generate(0, [])
    m = len(rows)

    # Build compatibility: two rows are compatible if they share no crack position
    compat = [[] for _ in range(m)]
    for i in range(m):
        for j in range(m):
            if rows[i].isdisjoint(rows[j]):
                compat[i].append(j)

    # DP: dp[i] = number of walls of current height ending with row i
    dp = [1] * m  # height 1

    for h in range(2, H + 1):
        ndp = [0] * m
        for i in range(m):
            for j in compat[i]:
                ndp[i] += dp[j]
        dp = ndp

    answer = sum(dp)
    print(answer)

if __name__ == "__main__":
    solve()
